File: android/app/src/main/java/com/mccontroller/core/LayoutSpec.kt
package com.mccontroller.core

/**
 * Position anchor for a widget. Maps to FrameLayout gravity flags.
 */
enum class Anchor {
    TopStart, TopCenter, TopEnd,
    CenterStart, CenterEnd,
    BottomStart, BottomCenter, BottomEnd,
}

/**
 * How the hotbar interprets left/right swipes.
 *
 * - [Precise]: a swipe lands on whatever slot the finger is over at the
 *   moment (original behavior). Drifting outside the strip clamps to the
 *   nearest extreme slot.
 * - [Relative]: scroll-wheel feel — direction matters, position doesn't.
 *   Every ~32dp of horizontal travel cycles the selection one slot in
 *   that direction, wrapping at slot 0 ↔ slot 8. Tap and long-press
 *   behaviors are unaffected.
 */
enum class HotbarSwipeMode { Precise, Relative }

/**
 * Per-widget layout: where it sits and how big it is.
 *
 * - `anchor` decides which screen edge(s) we measure from.
 * - `edgeMarginDp` is the horizontal distance from the start/end edge.
 * - `verticalMarginDp` is the distance from the top/bottom (or 0 for center).
 * - `widthDp` and `heightDp` of 0 mean "wrap_content" (used for LinearLayouts).
 */
data class WidgetSpec(
    val id: String,
    val anchor: Anchor,
    val edgeMarginDp: Float,
    val verticalMarginDp: Float,
    val widthDp: Float,
    val heightDp: Float,
)

/**
 * Per-mode layout: a map of widgets + global L/R margin offsets that
 * shift all start-anchored / end-anchored widgets uniformly (Honor-of-Kings
 * style "edge offset" sliders).
 */
data class ModeLayout(
    val widgets: Map<String, WidgetSpec>,
    val leftOffsetDp: Float = 0f,
    val rightOffsetDp: Float = 0f,
)

/**
 * Named layout profile: a complete layout for both in-game and UI modes.
 */
data class LayoutProfile(
    val name: String,
    val inGame: ModeLayout,
    val uiMode: ModeLayout,
    val hotbarSwipeMode: HotbarSwipeMode = HotbarSwipeMode.Precise,
)

/**
 * Hard-coded factory defaults — what the user sees the very first time
 * the app runs, mirroring the static positions in `activity_controller.xml`.
 *
 * Treat the LinearLayout containers (row_sneak_sprint, row_top_buttons,
 * column_ui_buttons) as single editable units — moving the row moves all
 * its children together. Sub-buttons inside keep their relative positions.
 */
object DefaultLayouts {

    val IN_GAME: ModeLayout = ModeLayout(
        widgets = mapOf(
            // Joystick activation zone (left side) — not editable in v3.
            "joystick" to WidgetSpec("joystick", Anchor.BottomStart, 0f, 0f, 360f, 280f),
            // Sneak alone above joystick (Sprint lives in the right-side fan)
            "btn_sneak" to WidgetSpec("btn_sneak", Anchor.BottomStart, 16f, 296f, 56f, 56f),

            // RIGHT-SIDE FAN (Honor-of-Kings style):
            //   LMB is the "attack" anchor — largest, deepest into the corner.
            //   RMB / Jump / Sprint fan out above-left along an arc of radius
            //   ~110dp from LMB's center, at angles 180°, 135°, 90°.
            "btn_lmb" to WidgetSpec("btn_lmb", Anchor.BottomEnd, 16f, 16f, 88f, 88f),
            "btn_rmb" to WidgetSpec("btn_rmb", Anchor.BottomEnd, 140f, 30f, 60f, 60f),
            "btn_jump" to WidgetSpec("btn_jump", Anchor.BottomEnd, 108f, 108f, 60f, 60f),
            "btn_sprint" to WidgetSpec("btn_sprint", Anchor.BottomEnd, 30f, 140f, 60f, 60f),

            // Top-right action buttons — individually editable.
            "btn_swap" to WidgetSpec("btn_swap", Anchor.TopEnd, 16f, 8f, 48f, 48f),
            "btn_inv" to WidgetSpec("btn_inv", Anchor.TopEnd, 72f, 8f, 48f, 48f),
            "btn_esc" to WidgetSpec("btn_esc", Anchor.TopEnd, 128f, 8f, 48f, 48f),

            // Hotbar — bottom-center, narrower (288dp = 9 slots × 32dp) so it
            // doesn't overlap with the right-side button fan. The slight overlap
            // with the joystick activation zone is fine since the joystick fades
            // when not in use.
            "hotbar" to WidgetSpec("hotbar", Anchor.BottomCenter, 0f, 8f, 288f, 40f),
        ),
    )

    val UI_MODE: ModeLayout = ModeLayout(
        widgets = mapOf(
            // Five individually-editable buttons stacked from the bottom-left,
            // sized to fit on a typical landscape screen (~400dp tall).
            "btn_ui_lmb" to WidgetSpec("btn_ui_lmb", Anchor.BottomStart, 24f, 20f, 72f, 72f),
            "btn_ui_rmb" to WidgetSpec("btn_ui_rmb", Anchor.BottomStart, 24f, 104f, 72f, 72f),
            "btn_ui_q" to WidgetSpec("btn_ui_q", Anchor.BottomStart, 24f, 188f, 56f, 56f),
            "btn_ui_shift" to WidgetSpec("btn_ui_shift", Anchor.BottomStart, 24f, 256f, 56f, 56f),
            "btn_ui_esc" to WidgetSpec("btn_ui_esc", Anchor.BottomStart, 24f, 324f, 56f, 56f),
        ),
    )

    val DEFAULT_PROFILE: LayoutProfile = LayoutProfile(
        name = "Default",
        inGame = IN_GAME,
        uiMode = UI_MODE,
    )

    /** Buttons that may have width/height edited (joystick excluded — v3 lock). */
    val RESIZABLE_IDS: Set<String> = setOf(
        "hotbar",
        "btn_sneak", "btn_sprint",
        "btn_lmb", "btn_rmb", "btn_jump",
        "btn_esc", "btn_inv", "btn_swap",
        "btn_ui_lmb", "btn_ui_rmb", "btn_ui_q", "btn_ui_shift", "btn_ui_esc",
    )

    /** All editable widget IDs, ordered for editor UI listing. */
    val IN_GAME_IDS: List<String> = listOf(
        "btn_sneak", "btn_sprint",
        "btn_lmb", "btn_rmb", "btn_jump",
        "btn_swap", "btn_inv", "btn_esc",
        "hotbar",
    )
    val UI_MODE_IDS: List<String> = listOf(
        "btn_ui_lmb", "btn_ui_rmb", "btn_ui_q", "btn_ui_shift", "btn_ui_esc",
    )
}

File: ios/McController/Core/LayoutSpec.swift
import Foundation
import CoreGraphics

/// Anchor — eight cardinal positions on the screen rectangle. Mirrors the
/// Android `Anchor` enum.
enum Anchor: String, Codable, CaseIterable, Sendable {
    case topStart, topCenter, topEnd
    case centerStart, centerEnd
    case bottomStart, bottomCenter, bottomEnd

    /// True when the anchor's horizontal position is determined relative to
    /// the canvas centre rather than an edge. On iOS we interpret `edge` on
    /// these anchors as a **signed offset from centre** (positive = right of
    /// centre, negative = left), so a centre-anchored widget can still be
    /// dragged horizontally.
    var isHorizontallyCentered: Bool {
        self == .topCenter || self == .bottomCenter
    }
}

/// Layout descriptor for a single widget. Sizes / margins are in points.
struct WidgetSpec: Codable, Hashable, Identifiable, Sendable {
    let id: String
    var anchor: Anchor
    /// For start / end anchors: distance from the start / end edge (≥ 0).
    /// For centre anchors: signed offset from the canvas centre.
    var edge: CGFloat
    /// Distance from the top / bottom (or below-centre for centre rows).
    var vertical: CGFloat
    var width: CGFloat       // 0 = wrap content
    var height: CGFloat
}

/// All the widget specs for one mode (InGame OR UiInteract).
struct ModeLayout: Codable, Hashable, Sendable {
    var widgets: [String: WidgetSpec]
    var leftOffset: CGFloat = 0
    var rightOffset: CGFloat = 0
}

enum HotbarSwipeMode: String, Codable, CaseIterable, Sendable {
    case precise        // position-based: slot = x / slotWidth
    case relative       // scroll-wheel: every N pt of travel cycles ±1
}

/// One named layout profile (the user can keep multiple).
struct LayoutProfile: Codable, Hashable, Sendable {
    var name: String
    var inGame: ModeLayout
    var uiMode: ModeLayout
    var hotbarSwipeMode: HotbarSwipeMode = .precise
}

// MARK: - Default layouts

enum DefaultLayouts {

    /// Widgets a user can resize / reposition.
    /// `joystick` is intentionally NOT here — its 360×280 activation zone is
    /// fixed in v3 (Android also locks it).
    static let resizableIDs: Set<String> = [
        "btn_sneak", "btn_lmb", "btn_rmb", "btn_jump", "btn_sprint",
        "btn_swap", "btn_inv", "btn_esc", "btn_close",
        "hotbar",
        "btn_ui_lmb", "btn_ui_rmb", "btn_ui_q", "btn_ui_shift", "btn_ui_esc",
    ]

    static let inGame = ModeLayout(widgets: [
        "joystick":    .init(id: "joystick",    anchor: .bottomStart,  edge: 0,   vertical: 0,   width: 360, height: 280),
        "btn_sneak":   .init(id: "btn_sneak",   anchor: .bottomStart,  edge: 16,  vertical: 296, width: 56,  height: 56),
        "btn_lmb":     .init(id: "btn_lmb",     anchor: .bottomEnd,    edge: 16,  vertical: 16,  width: 88,  height: 88),
        "btn_rmb":     .init(id: "btn_rmb",     anchor: .bottomEnd,    edge: 140, vertical: 30,  width: 60,  height: 60),
        "btn_jump":    .init(id: "btn_jump",    anchor: .bottomEnd,    edge: 108, vertical: 108, width: 60,  height: 60),
        "btn_sprint":  .init(id: "btn_sprint",  anchor: .bottomEnd,    edge: 30,  vertical: 140, width: 60,  height: 60),
        "btn_swap":    .init(id: "btn_swap",    anchor: .topEnd,       edge: 16,  vertical: 8,   width: 48,  height: 48),
        "btn_inv":     .init(id: "btn_inv",     anchor: .topEnd,       edge: 72,  vertical: 8,   width: 48,  height: 48),
        "btn_esc":     .init(id: "btn_esc",     anchor: .topEnd,       edge: 128, vertical: 8,   width: 48,  height: 48),
        // iOS-only: a movable close button that replaces the floating
        // top-right "back" chevron. Placed at top-start by default so it
        // doesn't collide with the swap / inv / esc row on the top-end.
        "btn_close":   .init(id: "btn_close",   anchor: .topStart,     edge: 16,  vertical: 8,   width: 48,  height: 48),
        "hotbar":      .init(id: "hotbar",      anchor: .bottomCenter, edge: 0,   vertical: 8,   width: 288, height: 40),
    ])

    static let uiMode = ModeLayout(widgets: [
        "btn_ui_lmb":   .init(id: "btn_ui_lmb",   anchor: .bottomStart, edge: 24, vertical: 20,  width: 72, height: 72),
        "btn_ui_rmb":   .init(id: "btn_ui_rmb",   anchor: .bottomStart, edge: 24, vertical: 104, width: 72, height: 72),
        "btn_ui_q":     .init(id: "btn_ui_q",     anchor: .bottomStart, edge: 24, vertical: 188, width: 56, height: 56),
        "btn_ui_shift": .init(id: "btn_ui_shift", anchor: .bottomStart, edge: 24, vertical: 256, width: 56, height: 56),
        "btn_ui_esc":   .init(id: "btn_ui_esc",   anchor: .bottomStart, edge: 24, vertical: 324, width: 56, height: 56),
        // iOS-only close button mirrors the in-game placement.
        "btn_close":    .init(id: "btn_close",    anchor: .topStart,    edge: 16, vertical: 8,   width: 48, height: 48),
    ])

    static let defaultProfile = LayoutProfile(
        name: "Default",
        inGame: inGame,
        uiMode: uiMode,
        hotbarSwipeMode: .precise
    )
}

// MARK: - Layout application

/// Translates a `WidgetSpec + ModeLayout` into a frame inside `bounds`.
/// Returns the frame in the same coordinate system as `bounds` (typically
/// the parent view's local space). Sizes of `0` mean "wrap content" — caller
/// must supply the intrinsic size.
enum LayoutApplier {

    static func frame(
        for spec: WidgetSpec,
        in mode: ModeLayout,
        bounds: CGRect,
        intrinsicSize: CGSize = .zero
    ) -> CGRect {
        let w = spec.width  > 0 ? spec.width  : intrinsicSize.width
        let h = spec.height > 0 ? spec.height : intrinsicSize.height

        let leftAdj  = mode.leftOffset
        let rightAdj = mode.rightOffset

        let x: CGFloat
        switch spec.anchor {
        case .topStart, .centerStart, .bottomStart:
            x = bounds.minX + spec.edge + leftAdj
        case .topEnd, .centerEnd, .bottomEnd:
            x = bounds.maxX - spec.edge - rightAdj - w
        case .topCenter, .bottomCenter:
            // `edge` is a signed offset from canvas centre (positive = right).
            x = bounds.midX - w / 2 + spec.edge
        }

        let y: CGFloat
        switch spec.anchor {
        case .topStart, .topCenter, .topEnd:
            y = bounds.minY + spec.vertical
        case .bottomStart, .bottomCenter, .bottomEnd:
            y = bounds.maxY - spec.vertical - h
        case .centerStart, .centerEnd:
            y = bounds.midY + spec.vertical
        }

        return CGRect(x: x, y: y, width: w, height: h)
    }
}
